src/transport/ExchangeMgr.h

```diff
@@ -0,0 +1,276 @@
+     * from the peer.
+     */
+    KeyErrorFunct OnKeyError;
+
+    void CancelRetrans();
+
+    /*
+     * in order to use reference counting (see refCount below)
+     * we use a hold/free paradigm where users of the exchange
+     * can hold onto it while it's out of their direct control
+     * to make sure it isn't closed before everyone's ready.
+     */
+    void AddRef();
+    void Close();
+    void Abort();
+    void Release();
+    CHIP_ERROR StartTimerT();
+
+private:
+    System::PacketBuffer * msg; // If we are re-transmitting, then this is the pointer to the message being retransmitted
+
+    // Trickle-controlled retransmissions:
+    uint32_t backoff;                        // backoff for sampling the numner of messages
+    uint8_t msgsReceived;                    // number of messages heard during the backoff period
+    uint8_t rebroadcastThreshold;            // re-broadcast threshold
+    BitFlags<uint16_t, ExFlagValues> mFlags; // Internal state flags
+
+    CHIP_ERROR ResendMessage();
+    bool MatchExchange(const PacketHeader & packetHeader, const PayloadHeader & payloadHeader);
+    static void CancelRetransmissionTimer(System::Layer * aSystemLayer, void * aAppState, System::Error aError);
+    static void TimerTau(System::Layer * aSystemLayer, void * aAppState, System::Error aError);
+    static void TimerT(System::Layer * aSystemLayer, void * aAppState, System::Error aError);
+
+    CHIP_ERROR StartResponseTimer();
+    void CancelResponseTimer();
+    static void HandleResponseTimeout(System::Layer * aSystemLayer, void * aAppState, System::Error aError);
+
+    uint32_t mPendingPeerAckId;
+    void DoClose(bool clearRetransTable);
+    CHIP_ERROR HandleMessage(const PacketHeader & packetHeader, const PayloadHeader & payloadHeader, System::PacketBuffer * msgBuf);
+    CHIP_ERROR HandleMessage(const PacketHeader & packetHeader, const PayloadHeader & payloadHeader, System::PacketBuffer * msgBuf,
+                             ExchangeContext::MessageReceiveFunct umhandler);
+
+    uint8_t mRefCount;
+};
+
+/**
+ *  @class ExchangeManager
+ *
+ *  @brief
+ *    This class is used to manage ExchangeContexts with other CHIP nodes.
+ *    It works on behalf of higher layers, creating ExchangeContexts and
+ *    handling the registration/unregistration of unsolicited message handlers.
+ *
+ */
+class DLL_EXPORT ExchangeManager : public SecureSessionMgrCallback
+{
+    friend class ExchangeContext;
+
+public:
+    enum State
+    {
+        kState_NotInitialized = 0, /**< Used to indicate that the ExchangeManager is not initialized */
+        kState_Initialized    = 1  /**< Used to indicate that the ExchangeManager is initialized */
+    };
+
+    ExchangeManager();
+    ExchangeManager(const ExchangeManager &) = delete;
+    ExchangeManager operator=(const ExchangeManager &) = delete;
+
+    SecureSessionMgrBase * MessageLayer; /**< [READ ONLY] The associated SecureSessionMgrBase object. */
```

Comment:
Do these fields (`MessageLayer`, and `State`) need to be public?

Reply:
Done.